src/morphodict/analysis/tag_map.py

```diff
@@ -0,0 +1,145 @@
+class UnknownTagError(KeyError):
+    """Raised when TagMap encounters an unknown tag"""
+
+
+class TagMap:
+    """Map between different sets of tags used by related FSTs
+
+    # Motivation
+
+    Sometimes different but related FSTs use similar sets of tags,
+    and they might want them in specific orders.
+
+    For example, one generic wordform analysis for ‘acâhkosa’ is
+    `acâhkosa+N+A+Der/Dim+N+A+Obv`. There’s a phrase-generation FST that
+    takes as input tags and definition in the form `Obv+Dim+ star`, and
+    outputs the inflected phrase `little star over there`. `Obv` has the
+    same tag name but is now a start tag, not an end tag, and `Der/Dim`
+    needs to be translated to just `Dim`. As well, the phrase-generation
+    FST has stricter ordering requirements on the input tags.
+
+    This class can handle all of that.
+
+    # Use
+
+    A TagMap is initialized with a sequence of (input_tag, output_tag,
+    precedence) tuples
+
+    input_tag can be:
+      - A literal tag, e.g., "N+", which will be matched exactly
+      - A tuple of tags, e.g., ("PV/e+, "+Ind") which will be matched as a
+        subsequence
+      - DEFAULT if this output_tag should be used if no other mapping
+      applies at this precedence level
+
+    output_tag can be:
+      - None if the input_tag should be dropped
+      - COPY_TAG_NAME if the characters of the input_tag match the
+        output_tag, for example: ("+Sg", COPY_TAG_NAME, x) means the same
+        thing as ("+Sg", "Sg+", x), but with less potential for copy-paste
+        mistakes.
+
+    All multi-mappings are applied before single maps, and consume their
+    tags. For example, a match on (("+A, "+B"), "foo", 1) will take the
+    tags "+A" and "+B" out of consideration before the rules ("+A",
+    COPY_TAG_NAME, 1) or ("+B", COPY_TAG_NAME, 1) are considered.
+
+    The precedence number is used to sort tags in the output list. For
+    example, if you want Pl/Sg before Px, you could give Pl and Sg
+    precedence number 1 and the possessives number 2. This precedence
+    number is associated with the output tag; it is an error to give a
+    different precedence value to multiple definitions that output the same
+    tag."""
+
+    DEFAULT = object()
+    COPY_TAG_NAME = object()
+
+    def __init__(self, *tag_definitions):
+        """
+        See the docs in crk_tag_map.py.
+        """
+        self._multi_mappings = []
+        self._tag_mapping = {}
+        self._precedences = {}
+        self._defaults_by_precedence = {}
+
+        for input_tag_spec, output_tag_spec, prec in tag_definitions:
+            if isinstance(input_tag_spec, tuple):
+                if output_tag_spec == TagMap.COPY_TAG_NAME:
+                    raise Exception(
+                        f"Error: cannot use copy with multi-tags: {input_tag_spec}"
```

Comment:
Is there a better exception class for this? Or can a more specific exception be made for this?